Allow specifying `token` parameter (SHA256 of key) to `/key/generate` API

msabramo opened the pull request:
This could be useful if you have multiple LiteLLM proxy instances or you are migrating from one to another and want to reuse existing key hashes. For example in the example below, John already has a key on some LiteLLM proxy that has a SHA256 hash of `60980d9c682a44cecdb4462f512b57e48fc63288680725962e87b9f764eae47a` and we want to enable that same key to work in a new LiteLLM proxy. We can do that by generating a key but specifying the `token` parameter to equal that existing hash:

```shell
litellm-proxy keys generate \
  --user-id=john \
  --team-id=team123 \
  --key-alias=john-key-1 \
  --key-name="Unknown key with hash 60980d9c682a44cecdb4462f512b57e48fc63288680725962e87b9f764eae47a" \
  --token=60980d9c682a44cecdb4462f512b57e48fc63288680725962e87b9f764eae47a```
```

The output of the above is:

```json
{
  "key_alias": "john-key-1",
  ...
  "user_id": "john",
  "team_id": "team123",
  ...
  "key": "(Unknown)",
  "token": "60980d9c682a44cecdb4462f512b57e48fc63288680725962e87b9f764eae47a",
  "key_name": "Unknown key with hash 60980d9c682a44cecdb4462f512b57e48fc63288680725962e87b9f764eae47a",
  ...
  "token_id": "60980d9c682a44cecdb4462f512b57e48fc63288680725962e87b9f764eae47a",
  ...
  "created_at": "2025-05-07T00:26:42.890000Z",
  "updated_at": "2025-05-07T00:26:42.890000Z"
}
```

## Relevant issues

<!-- e.g. "Fixes #000" -->

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [ ] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [ ] I have added a screenshot of my new test passing locally 
- [ ] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [ ] My PR's scope is as isolated as possible, it only solves 1 specific problem


## Type

<!-- Select the type of Pull Request -->
<!-- Keep only the necessary ones -->

🆕 New Feature

## Changes

msabramo committed:
Allow specifying key hash to /key/generate API

This allows one to set the key hash to some known existing key hash in
order to import keys from one LiteLLM proxy instance to another.

```shell
$ litellm-proxy http request POST /key/generate --json '{"token": "692b37538c7685a2c0529572d7b8eae87f990e572216a00f79baaf2db8895e08"}'
{
  ...
  "key": "(Unknown)",
  "token": "692b37538c7685a2c0529572d7b8eae87f990e572216a00f79baaf2db8895e08",
  ...
  "token_id": "692b37538c7685a2c0529572d7b8eae87f990e572216a00f79baaf2db8895e08",
  ...
}
```

msabramo committed:
Allow setting key_name

msabramo committed:
Add key generate tests

msabramo committed:
Add token & key_name to generate lib & CLI

```shell
$ litellm-proxy keys generate \
  --key-alias=my-key \
  --user-id=me@company.com \
  --token=2d7fd3dcfa6fb7a5142fcf234fc3e2e9dc983ad34ba435824929e8bb5ab28ebc \
  --key-name=some_key
{
  "key_alias": "my-key",
  ...
  "user_id": "me@company.com",
  ...
  "key": "(Unknown)",
  "token": "2d7fd3dcfa6fb7a5142fcf234fc3e2e9dc983ad34ba435824929e8bb5ab28ebc",
  "key_name": "some_key",
  ...
}
```

msabramo committed:
Update README.md files

ishaan-jaff commented on litellm/proxy/_types.py:
```diff
@@ -1593,12 +1595,13 @@ class UserAPIKeyAuth(
-                "api_key"
-            ).startswith("sk-"):
-                values.update({"api_key": hash_token(values.get("api_key"))})
+        api_key = values.get("api_key")
```
can you use `_hash_token_if_needed` in prox/utils.py here ?

msabramo replied:
Hmmm, just tried that and it resulted in a circular import so it's not a trivial change.

msabramo replied:
Well I think I can move `hash_token` and `_hash_token_if_needed` to `litellm.proxy._type` and then `utils.py` can have:

```python
from litellm.proxy._types import (
    hash_token,
    _hash_token_if_needed,
    ...
)
```

However, in order to test this with `make test-unit`, I will need #10662 to be merged...

msabramo replied:
Well, I'll just commit and push the change (3b2587a6e) and see if CI passes...

ishaan-jaff commented on litellm/proxy/client/cli/commands/keys.py:
```diff
@@ -88,6 +88,8 @@ def list(
 @click.option("--spend", type=float, help="Maximum spend limit for this key")
 @click.option("--duration", type=str, help="Duration for which the key is valid (e.g. '24h', '7d')")
 @click.option("--key-alias", type=str, help="Alias/name for the key")
+@click.option("--key-name", type=str, help="Display/abbreviation name for the key (separate from key_alias)")
```
can you add an example of this in your description eg. sk-12.....2oA

msabramo replied:
Yep, check out 7eb128b90.

ishaan-jaff commented on litellm/proxy/management_endpoints/key_management_endpoints.py:
```diff
@@ -568,6 +568,16 @@ async def generate_key_fn(  # noqa: PLR0915
+        elif "key" in data_json and data_json["key"] is not None:
+            data_json["token"] = hash_token(data_json["key"])
+        else:
+            data_json["token"] = f"sk-{secrets.token_urlsafe(LENGTH_OF_LITELLM_GENERATED_KEY)}"
```
it's not clear to me, why do we need this. I believe the code already generates the token if not provided

msabramo replied:
You're right! All of this crap can be removed! 00e96f77f

ishaan-jaff commented on litellm/proxy/management_endpoints/key_management_endpoints.py:
```diff
@@ -568,6 +568,16 @@ async def generate_key_fn(  # noqa: PLR0915
+        # --- TOKEN HANDLING LOGIC ---
+        if "token" in data_json and data_json["token"] is not None:
+            data_json["token"] = data_json["token"]
+        elif "key" in data_json and data_json["key"] is not None:
```
I believe our existing code already does this. I don't think this is necessary.

ishaan-jaff reviewed (changes requested):
reviewed

krrishdholakia commented:
Is this ready for review post-initial feedback?

msabramo committed:
Move hash_token and _hash_token_if_needed to _types.py

so that I can use `_hash_token_if_needed` in
`UserAPIKeyAuth.check_api_key` in `_types.py`.

addresses
[comment](https://github.com/BerriAI/litellm/pull/10601#discussion_r2076728733)
from @ishaan-jaff

msabramo committed:
Enhance keys generate docs with examples

addresses
[comment](https://github.com/BerriAI/litellm/pull/10601#discussion_r2076729055)
from @ishaan-jaff

msabramo committed:
Remove unnecessary code

addresses
[comment](https://github.com/BerriAI/litellm/pull/10601#discussion_r2076730535)
from @ishaan-jaff

msabramo committed:
Merge branch 'main' into allow-specifying-key-hash-to-key-generate-api

msabramo committed:
Fix lint errors

msabramo committed:
Fix more lint errors

msabramo committed:
More lint fixes

msabramo committed:
Fix more lint errors

msabramo committed:
Try to fix lint again for hash_token method

msabramo commented:
> Is this ready for review post-initial feedback?

This is ready now, as I addressed @ishaan-jaff's comments. Note the LiteLLM Mock Tests seem to be timing out after 8 minutes. I think merging #10484 will fix this.

msabramo commented:
Ready for review @ishaan-jaff
